### PR TITLE
#162 : add JacksonOption.RESPECT_JSONPROPERTY_REQUIRED

### PR DESCRIPTION
**Pull request by jjmbourdaret**
This PR will hopefully add an option, JacksonOption.RESPECT_JSONPROPERTY_REQUIRED, to populate "required" json schema property according to @JsonProperty(required...) value

### CODE REVIEW

**Comment by jjmbourdaret** [comment]
First ever PR....
I don't know if the test is up to the project standard, also the documentation accompanying might be a bit light .

File: jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonModule.java [hunk 1]
```diff
@@ -98,6 +98,10 @@ public void applyToConfigBuilder(SchemaGeneratorConfigBuilder builder) {
             generalConfigPart.withPropertySorter(new JsonPropertySorter(true));
         }
 
+        if (this.options.contains(JacksonOption.RESPECT_JSONPROPERTY_REQUIRED)) {
+            fieldConfigPart.withRequiredCheck( this::getRequiredCheckBasedOnJsonPropertyAnnotation );
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonModule.java:102]: Let's be consistent with existing similar code:
```suggestion
            fieldConfigPart.withRequiredCheck(this::getRequiredCheckBasedOnJsonPropertyAnnotation);
```

File: jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonModule.java [hunk 2]
```diff
@@ -248,4 +252,19 @@ protected boolean shouldIgnoreField(FieldScope field) {
         return beanDescription.findProperties().stream()
                 .noneMatch(propertyDefinition -> fieldName.equals(propertyDefinition.getInternalName()));
     }
+
+    /**
+     * Look-up the given field's {@link JsonProperty} annotation and its "required" attribute,  and calculate wether or not this field should be part of the required properties
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonModule.java:257]: two checkstyle rules complain here: the line is longer than 150 characters and a period symbol is missing to end the first sentence in the JavaDoc. Maybe just shorten it a bit:
```suggestion
     * Look-up the given field's {@link JsonProperty} annotation and consider its "required" attribute.
```

File: jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonModule.java [hunk 3]
```diff
@@ -248,4 +252,19 @@ protected boolean shouldIgnoreField(FieldScope field) {
...
+     *
+     * @param field field to look-up required strategy for
+     * @return whether the field should be in the "required" list or not
+     */
+    private boolean getRequiredCheckBasedOnJsonPropertyAnnotation(FieldScope field) {
+        JsonProperty jsonProperty = field.getAnnotationConsideringFieldAndGetter(JsonProperty.class) ;
+        if ( jsonProperty == null )
+            return false;
+        else
+            return jsonProperty.required();
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonModule.java:267]: checkstyle complains about missing curly braces. But we can shorten this a bit by getting rid of the `if`-`else` blocks:
```suggestion
        return jsonProperty != null && jsonProperty.required();
```

File: jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonOption.java [hunk 4]
```diff
@@ -63,5 +63,10 @@
     /**
      * Use this option to skip the transformation according to {@code @JsonTypeInfo} annotations (typically used to identify specific subtypes).
      */
-    IGNORE_TYPE_INFO_TRANSFORM;
+    IGNORE_TYPE_INFO_TRANSFORM,
+    /**
+     * Use this option to follow the required flag based on {@code @JsonProperty} required attribute
+     * By default, for backward compatibility, the "required" attribute of @JsonProperty is not
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JacksonOption.java:69]: checkstyle complains about a missing period symbol to end the first sentence.
The second part seems incomplete as well and isn't really necessary.
```suggestion
     * Use this option to include fields annotated with {@code @JsonProperty(required = true)} in the containing type's list of "required" properties.
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 5]
```diff
@@ -132,9 +132,9 @@ public void testApplyToConfigBuilderWithIgnoreTypeInfoTranformOption() {
 
     public Object[] parametersForTestApplyToConfigBuilderWithEnumOptions() {
         return new Object[][]{
-            {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE}},
-            {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY}},
-            {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE, JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY}}
+                {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE}},
+                {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY}},
+                {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE, JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY}}
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:137]: Unnecessary change
```suggestion
            {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE}},
            {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY}},
            {new JacksonOption[]{JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE, JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY}}
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 6]
```diff
@@ -171,12 +171,12 @@ private void verifyCommonConfigurations(boolean considerNamingStrategy) {
...
-            {"fieldWithSameValuePropertyAnnotation", null, "field-with-same-value-property-annotation"},
-            {"fieldWithNameOverride", "field override 1", "field-with-name-override"},
-            {"fieldWithNameOverrideOnGetter", "method override 1", "field-with-name-override-on-getter"},
-            {"fieldWithNameOverrideAndOnGetter", "field override 2", "field-with-name-override-and-on-getter"}
+                {"unannotatedField", null, "unannotated-field"},
+                {"fieldWithEmptyPropertyAnnotation", null, "field-with-empty-property-annotation"},
+                {"fieldWithSameValuePropertyAnnotation", null, "field-with-same-value-property-annotation"},
+                {"fieldWithNameOverride", "field override 1", "field-with-name-override"},
+                {"fieldWithNameOverrideOnGetter", "method override 1", "field-with-name-override-on-getter"},
+                {"fieldWithNameOverrideAndOnGetter", "field override 2", "field-with-name-override-and-on-getter"}
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:179]: unnecessary change
```suggestion
            {"unannotatedField", null, "unannotated-field"},
            {"fieldWithEmptyPropertyAnnotation", null, "field-with-empty-property-annotation"},
            {"fieldWithSameValuePropertyAnnotation", null, "field-with-same-value-property-annotation"},
            {"fieldWithNameOverride", "field override 1", "field-with-name-override"},
            {"fieldWithNameOverrideOnGetter", "method override 1", "field-with-name-override-on-getter"},
            {"fieldWithNameOverrideAndOnGetter", "field override 2", "field-with-name-override-and-on-getter"}
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 7]
```diff
@@ -198,11 +198,11 @@ public void testPropertyNameOverride(String fieldName, String expectedOverrideVa
...
-            {"unannotatedField", null},
-            {"fieldWithDescription", "field description 1"},
-            {"fieldWithDescriptionOnGetter", "getter description 1"},
-            {"fieldWithDescriptionAndOnGetter", "field description 2"},
-            {"fieldWithDescriptionOnType", null}
+                {"unannotatedField", null},
+                {"fieldWithDescription", "field description 1"},
+                {"fieldWithDescriptionOnGetter", "getter description 1"},
+                {"fieldWithDescriptionAndOnGetter", "field description 2"},
+                {"fieldWithDescriptionOnType", null}
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:205]: unnecessary change
```suggestion
            {"unannotatedField", null},
            {"fieldWithDescription", "field description 1"},
            {"fieldWithDescriptionOnGetter", "getter description 1"},
            {"fieldWithDescriptionAndOnGetter", "field description 2"},
            {"fieldWithDescriptionOnType", null}
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 8]
```diff
@@ -219,13 +219,36 @@ public void testDescriptionResolver(String fieldName, String expectedDescription
...
-            {"unannotatedField", null},
-            {"fieldWithDescription", null},
-            {"fieldWithDescriptionOnGetter", null},
-            {"fieldWithDescriptionAndOnGetter", null},
-            {"fieldWithDescriptionOnType", "class description text"}
+                {"unannotatedField", null},
+                {"fieldWithDescription", null},
+                {"fieldWithDescriptionOnGetter", null},
+                {"fieldWithDescriptionAndOnGetter", null},
+                {"fieldWithDescriptionOnType", "class description text"}
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:251]: unnecessary change
```suggestion
            {"unannotatedField", null},
            {"fieldWithDescription", null},
            {"fieldWithDescriptionOnGetter", null},
            {"fieldWithDescriptionAndOnGetter", null},
            {"fieldWithDescriptionOnType", "class description text"}
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 9]
```diff
@@ -219,13 +219,36 @@ public void testDescriptionResolver(String fieldName, String expectedDescription
...
     }
 
+    Object parametersForTestRequiredProperty() {
+        return new Object[][]{
+                {null,"requiredTrue", false},
+                {null,"requiredFalse", false},
+                {null,"requiredAbsent", false},
+                {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED,"requiredTrue", true},
+                {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED,"requiredFalse", false},
+                {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED,"requiredAbsent", false}
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:229]: inconsistent white-spaces; and we can add the `requiredDefault` case.
```suggestion
            {null, "requiredTrue", false},
            {null, "requiredFalse", false},
            {null, "requiredDefault", false},
            {null, "requiredAbsent", false},
            {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED, "requiredTrue", true},
            {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED, "requiredFalse", false},
            {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED, "requiredDefault", false},
            {JacksonOption.RESPECT_JSONPROPERTY_REQUIRED, "requiredAbsent", false}
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 10]
```diff
@@ -219,13 +219,36 @@ public void testDescriptionResolver(String fieldName, String expectedDescription
...
+        };
+    }
+
+    @Test
+    @Parameters
+    public void testRequiredProperty(JacksonOption requiredOption, String fieldName, boolean expectedRequired) {
+        new JacksonModule(requiredOption).applyToConfigBuilder(this.configBuilder);
+
+        FieldScope field = new TestType(TestClassWithRequiredAnnotatedFields.class).getMemberField(fieldName);
+        //this.fieldConfigPart.isRequired(field);
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:239]: remove commented-out code
```suggestion
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 11]
```diff
@@ -219,13 +219,36 @@ public void testDescriptionResolver(String fieldName, String expectedDescription
...
+
+        Assert.assertEquals(this.fieldConfigPart.isRequired(field), expectedRequired);
+    }
+
+
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:244]: one empty line is enough
```suggestion
```

File: jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java [hunk 12]
```diff
@@ -292,4 +315,16 @@ public Long getFieldWithDescriptionAndOnGetter() {
...
         }
     }
+
+    private static class TestClassWithRequiredAnnotatedFields {
+        @JsonProperty(required = true)
+        private String requiredTrue;
+
+        @JsonProperty(required = false)
+        private String requiredFalse;
+
```
Comment (CarstenWickner) [inline jsonschema-module-jackson/src/test/java/com/github/victools/jsonschema/module/jackson/JacksonModuleTest.java:325]: We can also include the case where the `required` field is not set, but the annotation is present.
```suggestion

        @JsonProperty
        private String requiredDefault;

```

File: slate-docs/source/includes/_jackson-module.md [hunk 13]
```diff
@@ -10,9 +10,9 @@ import com.github.victools.jsonschema.module.jackson.JacksonOption;
 
 JacksonModule module = new JacksonModule(
         JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE
-);
-SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2019_09)
-    .with(module);
+        );
+        SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2019_09)
+        .with(module);
```
Comment (CarstenWickner) [inline slate-docs/source/includes/_jackson-module.md:15]: unnecessary change
```suggestion
);
SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2019_09)
    .with(module);
```

File: slate-docs/source/includes/_jackson-module.md [hunk 14]
```diff
@@ -23,8 +23,9 @@ SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(Sc
 6. Treat enum types as plain strings as per the `@JsonValue` annotated method, if there is one and the `JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE` was provided (i.e. this is an "opt-in").
 7. Optionally: treat enum types as plain strings, as per each enum constant's `@JsonProperty` annotation, if all values of an enum have such annotations and the `JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY` was provided (i.e. this is an "opt-in").
 8. Optionally: sort an object's properties according to its `@JsonPropertyOrder` annotation, if the `JacksonOption.RESPECT_JSONPROPERTY_ORDER` was provided (i.e. this is an "opt-in").
```
Comment (CarstenWickner) [inline slate-docs/source/includes/_jackson-module.md:25]: I'd suggest mentioning the "required" setting above the more advanced topics.
```suggestion
6. Optionally: set a field as "required" as per `@JsonProperty` annotations, if the `JacksonOption.RESPECT_JSONPROPERTY_REQUIRED` was provided (i.e. this is an "opt-in").
7. Optionally: treat enum types as plain strings as per the `@JsonValue` annotated method, if there is one and the `JacksonOption.FLATTENED_ENUMS_FROM_JSONVALUE` was provided (i.e. this is an "opt-in").
8. Optionally: treat enum types as plain strings, as per each enum constant's `@JsonProperty` annotation, if all values of an enum have such annotations and the `JacksonOption.FLATTENED_ENUMS_FROM_JSONPROPERTY` was provided (i.e. this is an "opt-in").
9. Optionally: sort an object's properties according to its `@JsonPropertyOrder` annotation, if the `JacksonOption.RESPECT_JSONPROPERTY_ORDER` was provided (i.e. this is an "opt-in").
```

File: slate-docs/source/includes/_jackson-module.md [hunk 15]
```diff
@@ -23,8 +23,9 @@ SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(Sc
...
-9. Subtype resolution according to `@JsonSubTypes` on a supertype in general or directly on specific fields/methods as an override of the per-type behavior unless `JacksonOption.SKIP_SUBTYPE_LOOKUP` was provided (i.e. this is an "opt-out").
-10. Apply structural changes for subtypes according to `@JsonTypeInfo` on a supertype in general or directly on specific fields/methods as an override of the per-type behavior unless `JacksonOption.IGNORE_TYPE_INFO_TRANSFORM` was provided (i.e. this is an "opt-out").
+9. Optionally: set a field as "required" as per `@JsonProperty(required=... )` annotations, if the `JacksonOption.RESPECT_JSONPROPERTY_REQUIRED` was provided (i.e. this is an "opt-in").
```
Comment (CarstenWickner) [inline slate-docs/source/includes/_jackson-module.md:26]:
```suggestion
```

**Review by CarstenWickner — COMMENTED** [review]
Generally looking good. The only things missing are your entry as a new contributor in the main `pom.xml` and the mentioning of your contribution in the `CHANGELOG`, but I can add those for you.
Thanks a lot!
